pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -133,84 +129,172 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
 		names = append(names, name)
 	}
 
-	var certChain, keyPEM []byte
+	var certChain, keyPEM, caBundle []byte
 	var err error
 	if features.PilotCertProvider.Get() == KubernetesCAProvider {
 		log.Infof("Generating K8S-signed cert for %v", names)
 		certChain, keyPEM, _, err = chiron.GenKeyCertK8sCA(s.kubeClient.CertificatesV1beta1().CertificateSigningRequests(),
 			strings.Join(names, ","), hostnamePrefix+".csr.secret", namespace, defaultCACertPath)
-
-		s.caBundlePath = defaultCACertPath
+		if err != nil {
+			return fmt.Errorf("failed genrating ker cert by k8s: %v", err)
+		}
+		caBundle, err = ioutil.ReadFile(defaultCACertPath)
+		if err != nil {
+			return fmt.Errorf("failed reading %s: %v", defaultCACertPath, err)
+		}
 	} else if features.PilotCertProvider.Get() == IstiodCAProvider {
 		log.Infof("Generating istiod-signed cert for %v", names)
 		certChain, keyPEM, err = s.CA.GenKeyCert(names, SelfSignedCACertTTL.Get(), false)
+		if err != nil {
+			return fmt.Errorf("failed generating istiod key cert %v", err)
+		}
 
 		signingKeyFile := path.Join(LocalCertDir.Get(), "ca-key.pem")
 		// check if signing key file exists the cert dir
 		if _, err := os.Stat(signingKeyFile); err != nil {
 			log.Infof("No plugged-in cert at %v; self-signed cert is used", signingKeyFile)
-
```

Comment:
a silly question: not storing these DNS certs in the file-system is ok right? The file-system as of now appears to be backed up by a volume that is mounted (local-certs). Will this be removed in a later PR?

Reply:
> it appears this was mostly done for debugging purposes. can we have an info log printing out the entire DNS certificate to aid debugging?

Yes, could add a log

> The DNS file location as of now appears to be backed up by a volume that is mounted (local-certs). Will this be removed in a later PR?

Good question, with this, we can safely remove this volume mount.

Reply:
done